### ISSUE TITLE
App container unable to connect to network before sidecar is fully running

### ISSUE THREAD

**Issue by linsun** [istio/istio#11130]
**Describe the feature request**
We had users who spend very long time to debug why their app container stops working initially when sidecar is used in istio.  They have found out the app container could not reach out to network for simple things like clone a file from GitHub before the envoy proxy is ready and running.  It is hard to debug this because when they exec into the container after the deployment is running, everything works fine.

**Describe alternatives you've considered**
the current work group used by folks is to put a big sleep like 20 or 30 seconds in their app container to give enough time for envoy to start up.

This is fine once they discover the issue and understand how istio works better, but it can take days for them to discover the issue.

How can we make the experience better?
can we provide some startup hook so app container won't start till envoy sidecar is ready, if the app container starts very fast and requires network connectivity.

**Comment by linsun** [comment]
@esnible pls feel free to add things I missed.  cc @GregHanson

**Comment by esnible** [comment]
This may be a duplicate of https://github.com/istio/istio/issues/9454

**Comment by esnible** [comment]
This may be a duplicate of https://github.com/istio/istio/issues/4341

**Comment by jackkleeman** [comment]
Hey, we here at Monzo have open sourced our solution to this sequencing problem:
https://github.com/monzo/envoy-preflight
The idea is, it's a wrapper around your main application, which ensures it starts after envoy is live, and shuts down envoy when its done. You'll still need to prevent sigterms from reaching envoy.

**Comment by esnible** [comment]
Removing myself because I am not a sidecar networking guru.  That is what we need for this item.

**Comment by howardjohn** [comment]
Long term fix is https://github.com/istio/istio/issues/11366 or maybe https://github.com/kubernetes/kubernetes/issues/65502

**Comment by hzxuzhonghu** [comment]
As we have `ALLOW_ANY`, is this still a big problem?

**Comment by idouba** [comment]
Consider config postStart for app container to check envoy status.
`
 httpGet:
      path: /healthz/ready
`

**Comment by hzxuzhonghu** [comment]
I think this makes sense.
```
apiVersion: v1
kind: Pod
metadata:
  name: app
spec:
  containers:
  - name: app
    image: xxx
    command: 
    lifecycle:
      postStart:           # same as istio-proxy readiness probe, when this hook exec failed, the app container will be restarted
        httpGet:
          path: /healthz/ready
          port: 15020
```

**Comment by xiaozhongliu** [comment]
esnible's solution worked for us for a long period. Unfortunately the issue starts to occur again, and even worse.
Our external database can be unavailable for more than 8 seconds after the envoy is ready plus 5 seconds more sleep ...
```
until curl -s localhost:15000 > /dev/null; do echo '>>> Waiting for sidecar'; sleep 2 ; done ; echo '>>> Sidecar available'; sleep 5 ; ...
```

**Comment by Jonathan34** [comment]
It's been a year that this issue (or related issues) have been opened. 

It would be nice not to have the deployments know that they need to wait for the mesh' sidecar to be ready. That link should not exist and waiting for the sidecar is becoming a best practice and its a known common problem that leads to weak UX and onboarding of new users.

**Comment by howardjohn** [comment]
@Jonathan34 this has been around for a while because there is no easy solution. There are however two solutions in the Future

https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/sidecarcontainers.md will support this scenario natively in kubernetes. Basically it adds ordering to containers. There is an open PR for this in k8s so hopefully that lands in the near future. Once it is merged Istio will immediately support this.

The other potential option is to move envoy out of the pod (likely using CRI). This is a really complex change that requires a lot more work. See https://github.com/istio/istio/issues/11366 for more details.

We are very aware that this is not a great UX right now and hoping to improve it soon

**Comment by Jonathan34** [comment]
> @Jonathan34 this has been around for a while because there is no easy solution. There are however two solutions in the Future
> 
> https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/sidecarcontainers.md will support this scenario natively in kubernetes. Basically it adds ordering to containers. There is an open PR for this in k8s so hopefully that lands in the near future. Once it is merged Istio will immediately support this.
> 
> The other potential option is to move envoy out of the pod (likely using CRI). This is a really complex change that requires a lot more work. See #11366 for more details.
> 
> We are very aware that this is not a great UX right now and hoping to improve it soon

I appreciate the quick response. thanks. I will keep monitoring this issue

**Comment by luksa** [comment]
I've just submitted PR #24737 that fixes this issue. It's not ideal, but it works. I explain how it works here: https://medium.com/@marko.luksa/delaying-application-start-until-sidecar-is-ready-2ec2d21a7b74

**Comment by linsun** [comment]
@luksa very nice work!!  Will try this!

**Comment by christian-posta** [comment]
Nice @luksa !! This is something that has bothered me for a really long time! Thanks for writing up your findings in the blog post too.

**Comment by AntonySmirnoff** [comment]
now sure that I had the same issue, but this [holdApplicationUntilProxyStarts](https://istio.io/latest/news/releases/1.7.x/announcing-1.7/change-notes/) resolved mine.

**Comment by florianakos** [comment]
FYI - this may be useful for those out there running Istio older than v1.7 which has `holdApplicationUntilProxyStarts`.
So the snippet shared by **esnible** uses **localhost:15000** but this endpoint is for admin access which starts sooner than Envoy becomes operational, so for me using `localhost:15020/healthz/ready` instead helped a lot:

```
until [ $(curl --fail --silent --output /dev/stderr --write-out "%{http_code}" localhost:15020/healthz/ready) -eq 200 ]; do
  echo Waiting for proxy...
  sleep 1
done
```

**Comment by arocki7** [comment]
If the pod/app is very sensible to network connectivity at startup, I would recommend adding the following annotation to the application.

```yaml
annotations:
  proxy.istio.io/config: '{ "holdApplicationUntilProxyStarts": true }'
```

It would ensure that proxy is started up and functioning before starting the application. It solved the problem for me.

**Comment by omerfsen** [comment]
+1

**Comment by linsun** [comment]
Yes, this should be closed now I think! :)

**Comment by howardjohn** [comment]
I don't think this is done, it still doesn't support init containers or use stable APIs. Can we keep this open to track the long term solution?

**Comment by ldemailly** [comment]
This maybe a FAQ but how does `holdApplicationUntilProxyStarts` work? looking at the code it seems what it does is put the envoy proxy container first in the pod (instead of last); how is that holding the application? afaik containers aren't guaranteed to start sequentially (kubernetes/kubernetes#65502 is till open) or am I missing something? am I missing some probe interactions that achieve the goal of starting the app only after envoy is fully up and ready?

**Comment by howardjohn** [comment]
Containers do *start* sequentially - usually starting containers is nearly
instant though so it looks parallel. That issue you linked is about waiting
until previous ones are *ready*, not started.

The trick, which is really a hack relying on implementation details, is a
container is not considered "started" until the preStart hook (or maybe it
was postStart, I forget :-)) is finished. So the hook waits for envoy to be
ready.

**Comment by ldemailly** [comment]
Thanks a lot @howardjohn,  so if I understand correctly, istio-proxy pod does have a hook such as the readiness is disguised as an am-I-start'ed  hook (I didn't know about that hack that k8s was waiting for some feedback at all before starting the next pod; so I thought this was just changing/improving the race condition) - good to know it can be relied upon!

Any known downside ? (out of the kubectl exec thing which anyway seems to ask for a pod) - If no downside why isn't it the default? (I guess some app may be doing a lot of internal maybe disk based work at startup and thus don't need to wait for network/can benefit from parallel execution without waiting for pilot/envoy but that seems to be they could turn that on; most apps/services probably need to contact other systems during startup)

**Comment by mmerickel** [comment]
`kubectl run` is one scenario where you cannot select a container and if the hold is enabled then you cannot get into the application itself. You can opt-out via annotations in an override spec but this pretty tedious to do from the CLI. Changing which container is the "default" in a pod is not a great thing to do by default, but to be clear you can do it in the meshConfig on your cluster if you desire. Most apps I've run across do not need to talk to other services in the mesh as part of their startup but that depends on what you're up to I imagine.

**Comment by ldemailly** [comment]
it's not just in cluster it's all networking and it helps adoption to not have gotchas

**Comment by jdomag** [comment]
> Containers do *start* sequentially - usually starting containers is nearly instant though so it looks parallel. That issue you linked is about waiting until previous ones are *ready*, not started. The trick, which is really a hack relying on implementation details, is a container is not considered "started" until the preStart hook (or maybe it was postStart, I forget :-)) is finished. So the hook waits for envoy to be ready.

@howardjohn that's interesting - you probably mean postStart as per https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks
However after adding the `holdApplicationUntilProxyStarts: true` to istio config I can't see postStart hooks being added to istio-proxy container (there's only preStop that I used for another purpose) 

![image](https://user-images.githubusercontent.com/28705876/152511667-d6b0eb54-720c-4328-a27b-68508be2ea72.png)


Looking at the repository I found those lines:

https://github.com/istio/istio/blob/0faf5136361fa4cea96787872a024882dfc85d9d/operator/cmd/mesh/testdata/manifest-generate/output/sidecar_template.golden.yaml#L369

It doesn't make any sense for me - I used preStop and holdApplicationUntilProxyStarts   and ended up with preStop hook only. I understand though that both holdApplicationUntilProxyStarts and .Values.global.proxy.lifecycle.postStart can't be use the same time.

**Comment by mohammadsuha** [comment]
i am trying to do some setup by connecting to external service from init-container but i am unable connect to external service in ISTIO strict mode.
i found out ISTIO CNI so i want to know whether ISTIO CNI will help to resolve the issue?

https://istio.io/latest/docs/setup/additional-setup/cni/#compatibility-with-application-init-containers

in above document it says there are 3 ways we can do it

i want to know whether those 3 ways actually work to resolve the issue?

**Comment by sandeep-sharda-discovery** [comment]
Istio sidecar is killing the Postgress DB event listener connection frequently. We have used mentioned above workaround, 
    `proxy.istio.io/config: '{ "holdApplicationUntilProxyStarts": true }'
    traffic.sidecar.istio.io/excludeOutboundPorts: "5432"` 
however, the issue still occurs. Kindly update if any fix is available soon.

**Comment by ChimbuChinnadurai** [comment]
Another workaround is to run the init container as uid 1337
This will allow the initcontainer to bypass istio-proxy container for any external network access. 
```
securityContext:
  runAsUser: 1337
```

**Comment by hammadzz** [comment]
There really is no interim fix for this right? I tried setting `values.global.proxy.holdApplicationUntilProxyStarts=true` on istiod via it's helm release but it does not help. I see stackdriver export errors on almost every deploy of a new pod, which is quite annoying.

**Comment by luksa** [comment]
Kubernetes is getting proper sidecar support soon. Istio will be able to leverage this new feature to run the sidecar before any init container runs, which will enable network connectivity for init containers (through istio-proxy).

**Comment by omerfsen** [comment]
@luksa 
> Kubernetes is getting proper sidecar support soon. Istio will be able to leverage this new feature to run the sidecar before any init container runs, which will enable network connectivity for init containers (through istio-proxy).

Can you give me some link about this "Kubernetes is getting proper sidecar support soon" .. I want to read more about it..

**Comment by GregHanson** [comment]
There is some work on this in upstream istio already around this new Kubernetes feature:
 -  https://github.com/istio/istio/pull/44835
 - https://github.com/istio/istio/issues/44537

Related K8S links:
 - PR: https://github.com/kubernetes/kubernetes/pull/116429
 - KEP: https://github.com/kubernetes/enhancements/issues/753

**Comment by LukaszRacon** [comment]
If you want to eliminate sidecars - check ambient mesh:
https://istio.io/latest/blog/2022/introducing-ambient-mesh/

**Comment by GregHanson** [comment]
@LukaszRacon is right.  However, ambient was included in istio 1.18 but it is still considered [alpha feature status](https://istio.io/latest/docs/releases/feature-stages/) and isn't recommended for production use yet.  If you can afford to wait, ambient is definitely the way to go.

Granted the Sidecar KEP isn't implemented in istio yet and I don't know the timeline of when it will land

**Comment by keithmattix** [comment]
This is available via feature flag in #45959

**Comment by linsun** [comment]
closing this due to https://github.com/istio/istio/pull/45959

**Comment by thesuperzapper** [comment]
I just want to highlight something that I got confused by:

__THERE IS NO WAY TO USE THE ISTIO MESH IN `initContainers`__

This is because the `istio-proxy` sidecar must be running for the istio mesh to be available, and this will obviously not be the case during the init-container phase. 

Note, setting `holdApplicationUntilProxyStarts` to `true` will not fix this, because that setting only ensures your `containers` start after the `istio-proxy`.
